sshd-core/src/main/java/org/apache/sshd/common/util/SecurityUtils.java

```diff
@@ -89,6 +110,60 @@ public static boolean hasEcc() {
         return hasEcc;
     }
 
+    /**
+     * @return {@code true} if Diffie-Hellman Group Exchange is supported
+     * @see #getMaxDHGroupExchangeKeySize()
+     */
+    public static boolean isDHGroupExchangeSupported() {
+        return getMaxDHGroupExchangeKeySize() > 0;
+    }
+
+    /**
+     * @return The maximum supported Diffie-Hellman Group Exchange key size,
+     * or non-positive if not supported
+     */
+    public static int getMaxDHGroupExchangeKeySize() {
+        int maxSupportedKeySize;
+        synchronized (MAX_DHG_KEY_SIZE_HOLDER) {
+            maxSupportedKeySize = MAX_DHG_KEY_SIZE_HOLDER.get();
+            if (maxSupportedKeySize != 0) { // 1st time we are called ?
+                return maxSupportedKeySize;
+            }
+
+            if (System.getProperty(MAX_DHGEX_KEY_SIZE_PROP) != null) {
+                maxSupportedKeySize = Integer.parseInt(
+                        System.getProperty(MAX_DHGEX_KEY_SIZE_PROP));
+                if (!isDHGroupExchangeSupported(maxSupportedKeySize)) {
+                    maxSupportedKeySize = -1;   // signal the error and prevent further attempts
+                }
+            } else {
+                maxSupportedKeySize = -1;
+                for (int s = 1024; s <= 8192 && isDHGroupExchangeSupported(s); s += 1024) {
+                    maxSupportedKeySize = s;
+                }
+            }
+
+            MAX_DHG_KEY_SIZE_HOLDER.set(maxSupportedKeySize);
+        }
+
+        return maxSupportedKeySize;
+    }
```

Comment:
Then I will make the change, run it through the builder and if it is successful will merge into the master branch and let you know so you can close the pull request (we do not merge from them only use them for code discussions)